### ISSUE TITLE
implement a gradient descent on the homeostasis cost

### ISSUE THREAD

**Issue by laurentperrinet** [bicv/SHL_scripts#10]
the homeostasis using histogram equalization works and is great mathematically, but it slows down things... we could as well do a gradient descent of an homeostatic cost - hence have something similar as the implementation of the homeostasis in Olshausen, but on the probability of activation instead of the variance.

- one possible implementation is along that in Sandin et al 2016

- perhaps doable there in one line : https://github.com/bicv/SHL_scripts/blob/master/shl_scripts/shl_learn.py#L442

- then we need a way to tell to the algo to chose one method or the other (without making the code even more baroque, sigh...)

**Comment by laurentperrinet** [comment]
Victor has something working on his version. One issue is to know how to switch between all different homeostatic algorithms (olshausen / equalization / à-la-Sandin)

**Comment by slike04** [comment]
Nice, I did an implementation too...it's considerably faster
I get this after 10k iterations (15 mins running) on the classical 16x16 patches

<img width="493" alt="schermata 2017-12-05 alle 09 06 56" src="https://user-images.githubusercontent.com/29164075/33596347-ae9ad2c6-d99b-11e7-9792-73d5c3aabd8e.png">

And this after the usual 60k (1h30m running):

<img width="494" alt="schermata 2017-12-05 alle 09 08 41" src="https://user-images.githubusercontent.com/29164075/33596435-efb1dcc8-d99b-11e7-929b-20be5a4596cc.png">

It doesn't seem to converge on bigger patches tough...

**Comment by laurentperrinet** [comment]
@slike04  : can you push it ?

I will do the part to control which method we use...

**Comment by slike04** [comment]
I worked a bit on the code and I added the possibility of a dropout strategy on the coding phase and an exponential decay of the eta parameters;
20k iterations run in 15 minutes and I get this:
<img width="390" alt="screen shot 2017-12-05 at 18 09 44" src="https://user-images.githubusercontent.com/29164075/33620379-fefc0bf8-d9e7-11e7-90dd-8a9085cbd18d.png">
I will push before going home, don't know if it's a good strategy but for the the moment seems to work, I'll launch a simulation on bigger patches before going home

**Comment by slike04** [comment]
Never mind, the dropout was just a hype and makes no difference eheh, the exponential decay works fine tough

**Comment by laurentperrinet** [comment]
Do you use the same learning rule as @VictorBoutin ?

--
Laurent Perrinet - http://invibe.net

**Comment by slike04** [comment]
Not the same one, I think we're moving in similar directions...

**Comment by slike04** [comment]
UPDATE:
I didn't change much from the last version, the histogram equalization seems to work fine but the patches don't converge well for big dictionary size (and bigger l0_sparseness)
We should have a brainstorm about that

**Comment by laurentperrinet** [comment]
did you push that changes?

**Comment by slike04** [comment]
Yes it's still the one from Tuesday (?)...I think what I'm missing is a regularization on the sparse code, Naive errors...

**Comment by slike04** [comment]
Also zapping trough papers I found this one, it's not matching pursuit but maybe could be worth trying, seems easy to implement. 
[gregor-icml-10.pdf](https://github.com/bicv/SHL_scripts/files/1539117/gregor-icml-10.pdf)

**Comment by laurentperrinet** [comment]
implemented in https://github.com/bicv/SHL_scripts/commit/6769ab61f97baee3cbcdb205054c4650b2fdb0e6

**Comment by laurentperrinet** [comment]
note : to run, set ``do_HAP``to ``True`` and the parameter to scan is ``alpha_homeo`` - as before ``eta_homeo`` is the learning rate.

we should now check if the rule on https://github.com/bicv/SHL_scripts/blob/master/shl_scripts/shl_learn.py#L378 is the right one (and https://github.com/bicv/SHL_scripts/blob/master/shl_scripts/shl_learn.py#L379 is certainly not useful anymore)
